perf_benchmark/benchmark_main.cc

```diff
@@ -0,0 +1,482 @@
+#include "google/protobuf/text_format.h"
+#include "google/protobuf/struct.pb.h"
+#include "absl/strings/string_view.h"
+#include "absl/strings/str_format.h"
+#include "absl/strings/escaping.h"
+#include "absl/status/statusor.h"
+#include "absl/memory/memory.h"
+#include "google/api/service.pb.h"
+
+#include "perf_benchmark/utils.h"
+#include "perf_benchmark/benchmark_input_stream.h"
+#include "perf_benchmark/benchmark.pb.h"
+
+namespace google {
+namespace grpc {
+namespace transcoding {
+
+namespace perf_benchmark {
+namespace {
+using namespace benchmark;
+
+constexpr absl::string_view
+    kServiceConfigTextProtoFile = "benchmark_service.textproto";
+constexpr absl::string_view kNestedFieldName = "nested";
+constexpr absl::string_view kInnerMostNestedFieldName = "payload";
+constexpr absl::string_view kBytesPayloadMessageType = "BytesPayload";
+constexpr absl::string_view kStringPayloadMessageType = "StringPayload";
+constexpr absl::string_view kNestedPayloadMessageType = "NestedPayload";
+constexpr absl::string_view kInt32ArrayPayloadMessageType = "Int32ArrayPayload";
+constexpr absl::string_view
+    kStructPayloadMessageType = "google.protobuf.Struct";
+constexpr absl::string_view
+    kDoubleArrayPayloadMessageType = "DoubleArrayPayload";
+constexpr absl::string_view
+    kStringArrayPayloadMessageType = "StringArrayPayload";
+
+// Used for NestedPayload and StructPayload
+constexpr uint64_t kNumNestedLayersForStreaming = 32;
+// Used for ArrayPayload
+constexpr uint64_t kArrayPayloadLengthForStreaming = 1 << 10; // 1024
+// Used for BytesPayload
+constexpr uint64_t kBytesPayloadLengthForStreaming = 1 << 20; // 1 MiB
+// Used for Int32ArrayPayload
+constexpr uint64_t kInt32ArrayPayloadLengthForStreaming = 1 << 14; // 16384
+// Used for Segmented StringPayload
+constexpr uint64_t kSegmentedStringPayloadLength = 1 << 20; // 1 MiB
+}
+
+// Helper method to run Json Translation benchmark.
+//
+// Return absl::StatusOr<std::string> containing the parsed message if not
+// error.
+// state - ::benchmark::State& variable used for collecting metrics.
+// msg_type - Protobuf message name for translation.
+// json_msg - Complete input json message.
+// streaming - Flag for streaming testing. When true, a stream of `stream_size`
+//             number of `json_msg` will be fed into translation.
+// stream_size - Number of streaming messages.
+// chunk_per_msg - Number of data chunks per message.
+absl::StatusOr<std::string> BenchmarkJsonTranslation(::benchmark::State& state,
+                                                     absl::string_view msg_type,
+                                                     absl::string_view json_msg,
+                                                     bool streaming,
+                                                     uint64_t stream_size,
+                                                     uint64_t chunk_per_msg) {
+  // Load service config proto into Service object
+  google::api::Service service;
```

Comment:
Creating this `service` and `type_helper` is expensive. Since they stay the same across every single benchmark test, let's cache into a static variable? go/totw/110#the-fix-safe-initialization-no-destruction

Reply:
Done! Since `TypeHelper` is always built from the same Service config, I only added one global variable for `TypeHelper`.